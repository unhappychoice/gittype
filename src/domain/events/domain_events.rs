use super::Event;
use std::any::Any;

// Unified domain event enum for pattern matching
#[derive(Debug, Clone)]
pub enum DomainEvent {
    KeyPressed { key: char, position: usize },
    StageStarted { start_time: std::time::Instant },
    StagePaused,
    StageResumed,
    StageFinalized,
    StageSkipped,
    ChallengeLoaded { text: String, source_path: String },
}

impl Event for DomainEvent {
    fn as_any(&self) -> &dyn Any {
        self
    }
}
