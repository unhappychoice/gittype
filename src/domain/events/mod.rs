use std::any::{Any, TypeId};
use std::collections::HashMap;
use std::sync::{Arc, RwLock};

pub mod domain_events;

pub trait Event: Send + Sync + 'static {
    fn as_any(&self) -> &dyn Any;
}

pub trait EventHandler<E: Event>: Send + Sync {
    fn handle(&self, event: &E);
}

type BoxedHandler = Arc<dyn Fn(&dyn Event) + Send + Sync>;

pub struct EventBus {
    subscribers: Arc<RwLock<HashMap<TypeId, Vec<BoxedHandler>>>>,
}

impl EventBus {
    pub fn new() -> Self {
        Self {
            subscribers: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    pub fn publish<E: Event>(&self, event: E) {
        let type_id = TypeId::of::<E>();

        // Clone Arc<handlers> while holding the lock, then release it before calling them
        let handlers: Vec<BoxedHandler> = {
            let subscribers = self.subscribers.read().unwrap();
            let h = subscribers.get(&type_id)
                .map(|h| h.clone())
                .unwrap_or_default();
            h
        }; // Lock is released here

        // Call handlers without holding the lock
        for handler in handlers.iter() {
            handler(&event);
        }
    }

    pub fn subscribe<E: Event, F>(&self, handler: F)
    where
        F: Fn(&E) + Send + Sync + 'static,
    {
        let type_id = TypeId::of::<E>();
        let mut subscribers = self.subscribers.write().unwrap();

        let boxed_handler: BoxedHandler = Arc::new(move |event: &dyn Event| {
            if let Some(concrete_event) = event.as_any().downcast_ref::<E>() {
                handler(concrete_event);
            }
        });

        subscribers
            .entry(type_id)
            .or_insert_with(Vec::new)
            .push(boxed_handler);
    }
}

impl Default for EventBus {
    fn default() -> Self {
        Self::new()
    }
}

impl Clone for EventBus {
    fn clone(&self) -> Self {
        Self {
            subscribers: Arc::clone(&self.subscribers),
        }
    }
}
