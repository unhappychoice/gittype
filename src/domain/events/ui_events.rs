use super::Event;
use std::any::Any;

#[derive(Debug, Clone)]
pub struct RenderRequested {
    pub screen_name: String,
}

impl Event for RenderRequested {
    fn as_any(&self) -> &dyn Any {
        self
    }
}

#[derive(Debug, Clone)]
pub struct ScreenChanged {
    pub from: String,
    pub to: String,
}

impl Event for ScreenChanged {
    fn as_any(&self) -> &dyn Any {
        self
    }
}

#[derive(Debug, Clone)]
pub struct UserInputReceived {
    pub input: String,
}

impl Event for UserInputReceived {
    fn as_any(&self) -> &dyn Any {
        self
    }
}

#[derive(Debug, Clone)]
pub struct ConfigUpdated {
    pub key: String,
    pub value: String,
}

impl Event for ConfigUpdated {
    fn as_any(&self) -> &dyn Any {
        self
    }
}
