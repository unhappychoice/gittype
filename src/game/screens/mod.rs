pub mod animation_screen;
pub mod countdown_screen;
pub mod exit_summary_screen;
pub mod info_dialog;
pub mod loading_screen;
pub mod result_screen;
pub mod title_screen;
pub mod typing_screen;

pub use animation_screen::AnimationScreen;
pub use countdown_screen::CountdownScreen;
pub use exit_summary_screen::{ExitAction, ExitSummaryScreen};
pub use info_dialog::{InfoAction, InfoDialog};
pub use loading_screen::LoadingScreen;
pub use result_screen::{ResultAction, ResultScreen};
pub use title_screen::{TitleAction, TitleScreen};
pub use typing_screen::TypingScreen;
