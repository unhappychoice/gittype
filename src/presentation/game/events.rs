use std::any::Any;

// Re-export ScreenTransition as NavigateTo event
pub use super::models::screen::ScreenTransition as NavigateTo;

/// Event emitted when user requests to exit the application (Ctrl+C)
#[derive(Clone, Debug)]
pub struct ExitRequested;

impl crate::domain::events::Event for ExitRequested {
    fn as_any(&self) -> &dyn Any {
        self
    }
}
