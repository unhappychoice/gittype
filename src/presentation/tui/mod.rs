pub mod screen;
pub mod screen_manager;
pub mod screen_transition_manager;
pub mod screens;
pub mod views;

pub use screen::*;
pub use screen_manager::ScreenManager;
pub use screen_transition_manager::ScreenTransitionManager;
