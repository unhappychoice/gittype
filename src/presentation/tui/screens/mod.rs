// Game screens
pub mod analytics_screen;
pub mod animation_screen;
pub mod help_screen;
pub mod info_dialog;
pub mod loading_screen;
pub mod panic_screen;
pub mod records_screen;
pub mod session_detail_screen;
pub mod session_details_dialog;
pub mod session_failure_screen;
pub mod session_summary_screen;
pub mod session_summary_share_screen;
pub mod settings_screen;
pub mod stage_summary_screen;
pub mod title_screen;
pub mod total_summary_screen;
pub mod total_summary_share_screen;
pub mod typing_screen;
pub mod version_check_screen;

// CLI screens
pub mod repo_list_screen;
pub mod repo_play_screen;
pub mod trending_language_selection_screen;
pub mod trending_repository_selection_screen;

// Re-exports
pub use analytics_screen::AnalyticsScreen;
pub use animation_screen::AnimationScreen;
pub use help_screen::HelpScreen;
pub use info_dialog::{InfoDialogScreen, InfoDialogScreenDataProvider};
pub use loading_screen::LoadingScreen;
pub use panic_screen::PanicScreen;
pub use records_screen::RecordsScreen;
pub use repo_list_screen::{RepoListScreen, RepoListScreenDataProvider};
pub use repo_play_screen::{RepoPlayScreen, RepoPlayScreenDataProvider};
pub use session_detail_screen::SessionDetailScreen;
pub use session_details_dialog::SessionDetailsDialog;
pub use session_failure_screen::SessionFailureScreen;
pub use session_summary_screen::{ResultAction, SessionSummaryScreen};
pub use session_summary_share_screen::SessionSummaryShareScreen;
pub use settings_screen::SettingsScreen;
pub use stage_summary_screen::StageSummaryScreen;
pub use title_screen::{TitleAction, TitleScreen};
pub use total_summary_screen::TotalSummaryScreen;
pub use total_summary_share_screen::TotalSummaryShareScreen;
pub use trending_language_selection_screen::{
    TrendingLanguageSelectionScreen, TrendingLanguageSelectionScreenDataProvider,
};
pub use trending_repository_selection_screen::{
    TrendingRepositorySelectionScreen, TrendingRepositorySelectionScreenDataProvider,
};
pub use typing_screen::TypingScreen;
pub use version_check_screen::{VersionCheckResult, VersionCheckScreen};
