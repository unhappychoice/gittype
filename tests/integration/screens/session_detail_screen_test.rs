use crate::integration::screens::mocks::records_screen_mock::MockRecordsDataProvider;
use crate::integration::screens::mocks::session_repository_mock::MockSessionRepository;
use crossterm::event::{KeyCode, KeyEvent, KeyModifiers};
use gittype::domain::events::EventBus;
use gittype::presentation::game::events::NavigateTo;
use gittype::presentation::tui::ScreenDataProvider;
use gittype::presentation::tui::screens::{RecordsScreen, SessionDetailScreen};
use gittype::presentation::tui::Screen;
use std::sync::{Arc, Mutex};

// Helper function to create and initialize SessionDetailScreen from RecordsScreen
fn create_initialized_session_detail_screen(event_bus: EventBus) -> SessionDetailScreen {
    let mut screen =
        SessionDetailScreen::new(event_bus).with_session_repository(MockSessionRepository::new());

    let mut records = RecordsScreen::new(EventBus::new());
    let data = MockRecordsDataProvider.provide().unwrap();
    records.init_with_data(data).unwrap();
    records.set_selected_session_from_index(0);
    screen.on_pushed_from(&records).unwrap();

    screen
}

screen_snapshot_test!(
    test_session_detail_screen_snapshot,
    SessionDetailScreen,
    SessionDetailScreen::new(EventBus::new()).with_session_repository(MockSessionRepository::new()),
    pushed_from = {
        let mut records = RecordsScreen::new(EventBus::new());
        let data = MockRecordsDataProvider.provide().unwrap();
        records.init_with_data(data).unwrap();
        records.set_selected_session_from_index(0);
        records
    }
);

// Event-producing key tests
#[test]
fn test_session_detail_screen_esc_pops() {
    let event_bus = EventBus::new();
    let events = Arc::new(Mutex::new(Vec::new()));
    let events_clone = Arc::clone(&events);

    event_bus.subscribe(move |event: &NavigateTo| {
        events_clone.lock().unwrap().push(event.clone());
    });

    let mut screen = create_initialized_session_detail_screen(event_bus);

    screen
        .handle_key_event(KeyEvent::new(KeyCode::Esc, KeyModifiers::empty()))
        .unwrap();

    let captured_events = events.lock().unwrap();
    assert_eq!(captured_events.len(), 1);
}

#[test]
fn test_session_detail_screen_ctrl_c_exits() {
    let event_bus = EventBus::new();
    let events = Arc::new(Mutex::new(Vec::new()));
    let events_clone = Arc::clone(&events);

    event_bus.subscribe(move |event: &NavigateTo| {
        events_clone.lock().unwrap().push(event.clone());
    });

    let mut screen = create_initialized_session_detail_screen(event_bus);

    screen
        .handle_key_event(KeyEvent::new(KeyCode::Char('c'), KeyModifiers::CONTROL))
        .unwrap();

    let captured_events = events.lock().unwrap();
    assert_eq!(captured_events.len(), 1);
}

// Non-event key tests
#[test]
fn test_session_detail_screen_up_scrolls() {
    let event_bus = EventBus::new();
    let mut screen = create_initialized_session_detail_screen(event_bus);

    // Should not panic
    screen
        .handle_key_event(KeyEvent::new(KeyCode::Up, KeyModifiers::empty()))
        .unwrap();
}

#[test]
fn test_session_detail_screen_down_scrolls() {
    let event_bus = EventBus::new();
    let mut screen = create_initialized_session_detail_screen(event_bus);

    // Should not panic
    screen
        .handle_key_event(KeyEvent::new(KeyCode::Down, KeyModifiers::empty()))
        .unwrap();
}
