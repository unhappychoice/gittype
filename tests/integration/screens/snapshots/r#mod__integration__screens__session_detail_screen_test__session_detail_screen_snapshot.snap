---
source: tests/integration/screens/session_detail_screen_test.rs
expression: output
---
Session Details                                                                                                         
┌Session───────────────────────────────────────────────────┐┌Performance───────────────────────────────────────────────┐
│                                                          ││                                                          │
│  Repository: unhappychoice/gittype                       ││  Tier/Rank: Gold/Advanced                                │
│  Started: 2024-10-07 12:30:00                            ││  Score: 1200.0                                           │
│  Branch: main                                            ││  CPM: 375.0                                              │
│  Commit: abc123                                          ││  WPM: 75.0                                               │
│                                                          ││  Accuracy: 96.0%                                         │
│                                                          ││  Duration: 1m 0s                                         │
│                                                          ││  Completed Stage: 3/3                                    │
│                                                          ││                                                          │
│                                                          ││                                                          │
└──────────────────────────────────────────────────────────┘└──────────────────────────────────────────────────────────┘
┌Stage Details (3 stages)──────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                                                                                      │
│  Stage #1 [COMPLETED]                                                                                                │
│    File: src/main.rs:1-20                                                                                            │
│    Score: 400.0  CPM: 375.0    WPM: 75.0                                                                             │
│    Keystrokes: 150  Mistakes: 5  Accuracy: 96.7%  Duration: 20000ms                                                  │
│                                                                                                                      │
│  Stage #2 [COMPLETED]                                                                                                │
│    File: src/lib.rs:10-30                                                                                            │
│    Score: 380.0  CPM: 360.0    WPM: 72.0                                                                             │
│    Keystrokes: 175  Mistakes: 8  Accuracy: 95.4%  Duration: 22000ms                                                  │
│                                                                                                                      │
│  Stage #3 [COMPLETED]                                                                                                │
│    File: src/domain/mod.rs:5-25                                                                                      │
│    Score: 420.0  CPM: 390.0    WPM: 78.0                                                                             │
│    Keystrokes: 175  Mistakes: 7  Accuracy: 96.0%  Duration: 18000ms                                                  │
│                                                                                                                      │
│                                                                                                                      │
│                                                                                                                      │
│                                                                                                                      │
│                                                                                                                      │
│                                                                                                                      │
│                                                                                                                      │
│                                                                                                                      │
│                                                                                                                      │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
                                            [↑↓/JK] Scroll Stages  [ESC] Back
